Publish blog post (#1)
greg0ire/wiremock-blogpost: index.md

# Wiremock + testcontainers + Algolia + Go = ❤️

> 

Reviewer: Consider adding a brief overview of what the article will cover and/or a table of contents.

Author: Addressed at the end of the first section.

> When dealing with a SaaS like Algolia, testing can be awkward. Ideally, you

Reviewer: Effective intro that sets the stage by explaining the challenges and the need for a better solution 🙌

that seems to be the next best thing for the job: [Wiremock](https://wiremock.org).

Wiremock is a testing tool that comes with a so-called ["record and
playback"](https://wiremock.org/docs/record-playback) feature.

It means you can do this once in your local environment:

```
┌────────────┐          ┌────────────────────────────┐       ┌─────────┐
│            ├─────────►│                            ├──────►│         │
│Your service│          │ Wiremock in recording mode │       │ Algolia │
│            │◄─────────┤                            │◄──────┤         │
└────────────┘          └────────────────────────────┘       └─────────┘
```

In recording mode, you give Wiremock a URL to record, and it will store files
representing the requests you made, and the corresponding responses. With
Algolia, it can be quite long, especially if you [wait for
operations](https://www.algolia.com/doc/api-reference/api-methods/wait-task).
What happens in practice is that the SDK will use a polling mechanism to check
if your task is published. This will result in a lot of similarly looking files.
This is not very interesting to reproduce in your test, so I recommend simply

> deleting files representing a negative response. The files have names that are

Reviewer: The term "negative response" could be clarified with an example.

> a bit ugly, so I usually rename them for clarity.

Reviewer: Providing examples and renaming suggestions for file names would be helpful:

```suggestion
deleting files representing a negative response. The files have names that are
a bit ugly, so I usually rename them for clarity. 
For example, you might rename 
`1_indexes_test-index_task_226434943725-6e8689fa-9bbb-43fb-9d24-6824c02fc7d5.json`
to `index_test_task_published.json`.
```

that's not a great solution because Gitlab services do not expose the full
power of Docker. For instance, mounting a volume is not possible, probably not
without heavy involvement of privileged users.

A great alternative is [testcontainers](https://testcontainers.com) +
[testcontainers Cloud](https://testcontainers.com/cloud/). Testcontainers is a
library available in many languages that allows you to start and stop Docker
containers during your tests, making it possible to get good isolation between
 tests.
Testcontainers Cloud is a service that allows you to run start said containers
on a remote infrastructure, which means you do not need to do Docker in Docker
in your CI.
Locally, you would still use a local docker container, but in the CI,
tescontainers will send requests to testcontainers cloud, to start and stop
containers. Enough unpaid endorsement, let's get to the code.

## Demo time

Let us follow [Algolia's quickstart guide for
Golang](https://www.algolia.com/doc/libraries/go/v4/) and see how easy it is to
test.

If you want to follow along, install
[mise-en-place](https://mise.jdx.dev/getting-started.html) and let's go!

### Installing Go

```console

> mise use go@1.24

Reviewer: In addition to installing / setting up the version of Go, this section could benefit from more detailed project setup instruction. For example, including commands like `go mod init module/path/indexer`.

```

### Installing the Algolia SDK

```console
go get github.com/algolia/algoliasearch-client-go/v4/search
```

### Setting up the environment

At this point, you will need to set up a test Algolia application. Once you are
done, you should have an application ID and an API key.

Let us use an unversioned env file to store our credentials.

```toml
# mise.toml
_.file = ".env"
```

```shell
# .env
ALGOLIA_APP_ID=your_app_id

> ALGOLIA_API_KEY=your_app_key

Reviewer: I'd advise clarifying that environment variable values need to be replaced with valid Algolia API ID/key sets.

```

```shell
# .gitignore
/.env
```

> ### Writing the code under test

Reviewer: NIT: This could be clearer:

```suggestion
### Writing the code to be tested
```

Author: It was a reference to https://en.wikipedia.org/wiki/System_under_test, but you're right, let's reword this. The reference does not bring much value.

Let us take the code from Algolia's quickstart guide and split it into two files:

First, we have the code under test where the only changes are getting the
environment variables from the actual environment, and renaming packages and
functions.

```go

> package indexer

Reviewer: Specifying each `.go` file path and filename would be helpful.

import (
	"os"

	"github.com/algolia/algoliasearch-client-go/v4/algolia/search"
)

func indexRecord() {
	appID := os.Getenv("ALGOLIA_APP_ID")
	apiKey := os.Getenv("ALGOLIA_API_KEY")
	indexName := "test-index"

	record := map[string]any{
		"objectID": "object-1",
		"name":     "test record",
	}

	client, err := search.NewClient(appID, apiKey)

	if err != nil {
		panic(err)
	}

	// Add record to an index
	saveResp, err := client.SaveObject(
		client.NewApiSaveObjectRequest(indexName, record),
	)

	if err != nil {
		panic(err)
	}

	// Wait until indexing is done
	_, err = client.WaitForTask(indexName, saveResp.TaskID)

	if err != nil {
		panic(err)
	}
}
```

That call to `WaitForTask` is what is going to take the most time, and a good
reason not to use a real Algolia instance in your test suite. That's what we
are going to try first though.

### Writing the test with a real Algolia instance

Let's start simple and write a first version of the test that talks directly to
Algolia:

```go

> package indexer

Reviewer: Same suggestion regarding filename and path.

import (
	"os"
	"testing"

	"github.com/algolia/algoliasearch-client-go/v4/algolia/debug"
	"github.com/algolia/algoliasearch-client-go/v4/algolia/search"
)

> func TestIndexRecord(t *testing.T) {

Reviewer: Adding comments within the test code snippets to explain key parts of the code would make it easier for readers to understand what each part does 🙏

Author: I tried adding a few, if you see more places that lack comments, I'm interested.

Reviewer: Added another comment with suggestions, resolving.

FAIL    algolia-wiremock-testcontainers 185.745s
FAIL
```

I have many applications on this Metis instance, some of which are very busy,
let us patch that real quick:

```
// Wait until indexing is done
_, err = client.WaitForTask(
	indexName,
	saveResp.TaskID,
	search.WithMaxRetries(100),
)
```

Exactly the type of thing that unit tests will not catch.

After that, the test passes (but it took several minutes to run).
Great! Now, let's add a proxy in the middle, and record all this.

### Adding Wiremock in record mode 📼

We are using Docker, so if we want to obtain the so-called
"mapping files" wiremock will create, we need to mount a volume on our Docker
container, and mount it in the right location.

Let us add 2 new dependencies to our project:

We could interact with Wiremock by calling the REST API with the `net/http`
package, but as it turns out, there is a dedicated SDK for that, and it supports
recording since [this pull request I sent](https://github.com/wiremock/go-wiremock/pull/33).

At the time of writing, the PR is merged but not released yet, so for now,
let's use a commit hash:

```console

> go get github.com/testcontainers/testcontainers-go@553afd3c51228cedff0fbbb744b6da94456dbfe1

Reviewer: The mentioned merged PR is in `github.com/wiremock/go-wiremock` but instructions to retrieve an untagged package version from a commit hash for `github.com/testcontainers/testcontainers-go` instead. Please clarify this discrepancy (it looks as indeed it should be go-wiremock).

Author: You're absolutely right, good catch!

```

Next, we will need a way to start and stop the Wiremock container, and for that
as well, there is a library:

```console
go get github.com/wiremock/wiremock-testcontainers-go@v1.0.0-alpha-10
```

Yes, this is alpha software 😬

Let us start the container, with a volume mounting `testdata` in the current
directory on `/home/wiremock/mappings` in the container. This is where Wiremock
will create json files.

```go

> // indexer_test.go

Reviewer: It would be helpful to specify where the new pieces of updated code should go in each file. Consider including entire files with imports or providing a GitHub repo with commits for each stage of the demo.

Author: This PR contains all the files, but you're right, I will add a header linking to it.

Author: Currently in the process of doing the latter, and it helps me find issues in the blogpost. Thanks for this suggestion!

        state:                                                     | state: Started]
        scenario-1-1-indexes-test-index-task-226434943725-43]      |
                                                                   |
        -----------------------------------------------------------------------------------------------------------------------
```

After dropping `"requiredScenarioState" : "scenario-1-1-indexes-test-index-task-226434943725-43",`
from the mapping file about polling, the test passes again, only this time, it
passes in under 2 seconds.

### Making it work in the CI

After pushing the code, I got a bad surprise: the test fails in the CI, with
the following message:

```
tc-wiremock.go:73: create container: container create: Error response from daemon: Invalid bind mount config: mount source "/builds/product-discovery/ms.indexer/internal/import/brandsuggestion/testdata" is forbidden by the allow list [/home /tmp] - update the bind mounts configuration and restart the agent to enable
```

It would seem that we cannot use a bind mount in the CI. Let us use our

> `record` constant to make the container options conditional:

Reviewer: It would be interesting to have an explanation as to why the bind mount configuration is an issue in the CI environment and how the solution addresses this problem.

Author: I must admit the cause is not entirely clear to me; searching for the error message on DuckDuckGo, Google or Github yields nothing, and I suspect it might come from Testcontainers could own configuration. Looking at [their docs](https://testcontainers.com/cloud/docs/#mounting-the-local-filesystem), I get even more confused because it states that mounting into countainers is not even implemented.

I will however elaborate on why the solution addresses the problem.

// indexer_test.go

package indexer

import (
	"os"
	"testing"

	"github.com/algolia/algoliasearch-client-go/v4/algolia/debug"
	"github.com/algolia/algoliasearch-client-go/v4/algolia/search"
)

func TestIndexRecord(t *testing.T) {
	debug.Enable() // helps with seeing the progress, since this is super long

	appID := os.Getenv("ALGOLIA_APP_ID")
	apiKey := os.Getenv("ALGOLIA_API_KEY")
	indexName := "test-index"

	client, err := search.NewClient(appID, apiKey)
	if err != nil {
		t.Fatalf("Failed to create client: %v", err)
	}

	// Create index indirectly by setting settings
	_, err = client.SetSettings(client.NewApiSetSettingsRequest(
		"test-index",
		search.NewEmptyIndexSettings().SetSearchableAttributes([]string{"name"}),
	))

	if err != nil {
		t.Fatalf("Failed to set settings: %v", err)
	}

	t.Cleanup(func() {
		_, err := client.DeleteIndex(client.NewApiDeleteIndexRequest(indexName))
		if err != nil {
			t.Fatalf("Failed to delete index: %v", err)
		}
	})

	indexRecord()

	// Search for 'test'
	searchResp, err := client.Search(
		client.NewApiSearchRequest(
			search.NewEmptySearchMethodParams().SetRequests(
				[]search.SearchQuery{
					*search.SearchForHitsAsSearchQuery(

> 						search.NewEmptySearchForHits().SetIndexName(indexName).SetQuery("time"),

Author: Big blunder both I and Algolia need to fix.

```suggestion
						search.NewEmptySearchForHits().SetIndexName(indexName).SetQuery("test"),
```

		"name":     "test record",
	}

	client, err := search.NewClient(appID, apiKey)

	if err != nil {
		panic(err)
	}

	// Add record to an index
	saveResp, err := client.SaveObject(
		client.NewApiSaveObjectRequest(indexName, record),
	)

	if err != nil {
		panic(err)
	}

	// Wait until indexing is done
	_, err = client.WaitForTask(indexName, saveResp.TaskID)

	if err != nil {
		panic(err)
	}
}
```

To make it work, you will need to install the Algolia SDK:

```console
go get github.com/algolia/algoliasearch-client-go/v4
go mod tidy
```

That call to `WaitForTask` is what is going to take the most time, and a good
reason not to use a real Algolia instance in your test suite. That's what we
are going to try first though.

### Writing the test with a real Algolia instance

Let's start simple and write a first version of the test that talks directly to
Algolia:

```go
// indexer_test.go

package indexer

import (
	"os"
	"testing"

	"github.com/algolia/algoliasearch-client-go/v4/algolia/debug"
	"github.com/algolia/algoliasearch-client-go/v4/algolia/search"
)

func TestIndexRecord(t *testing.T) {
	debug.Enable() // helps with seeing the progress, since this is super long

> 	appID := os.Getenv("ALGOLIA_APP_ID")
> 	apiKey := os.Getenv("ALGOLIA_API_KEY")
> 	indexName := "test-index"
> 
> 	client, err := search.NewClient(appID, apiKey)
> 	if err != nil {
> 		t.Fatalf("Failed to create client: %v", err)
> 	}
> 
> 	// Create index indirectly by setting settings
> 	_, err = client.SetSettings(client.NewApiSetSettingsRequest(
> 		"test-index",
> 		search.NewEmptyIndexSettings().SetSearchableAttributes([]string{"name"}),
> 	))
> 
> 	if err != nil {
> 		t.Fatalf("Failed to set settings: %v", err)
> 	}
> 
> 	t.Cleanup(func() {
> 		_, err := client.DeleteIndex(client.NewApiDeleteIndexRequest(indexName))
> 		if err != nil {
> 			t.Fatalf("Failed to delete index: %v", err)
> 		}
> 	})
> 
> 	indexRecord()
> 
> 	// Search for 'test'
> 	searchResp, err := client.Search(

Reviewer: Regarding my previous suggestion to add more comments on the test code snippets: even if might feel redundant given the relatively short length of the snippet, I find that it adds readability to over-explain everything that is done, such as: 

```suggestion
        // Get Algolia credentials from environment variables
	appID := os.Getenv("ALGOLIA_APP_ID")
	apiKey := os.Getenv("ALGOLIA_API_KEY")
	indexName := "test-index"

        // Create a new Algolia client
	client, err := search.NewClient(appID, apiKey)
	if err != nil {
		t.Fatalf("Failed to create client: %v", err)
	}

	// Create index indirectly by setting settings
	_, err = client.SetSettings(client.NewApiSetSettingsRequest(
		"test-index",
		search.NewEmptyIndexSettings().SetSearchableAttributes([]string{"name"}),
	))
	if err != nil {
		t.Fatalf("Failed to set settings: %v", err)
	}

        // Ensure the test index is deleted after the test completes
	t.Cleanup(func() {
		_, err := client.DeleteIndex(client.NewApiDeleteIndexRequest(indexName))
		if err != nil {
			t.Fatalf("Failed to delete index: %v", err)
		}
	})

        // Call the function under test - this should index a record in Algolia
	indexRecord()

	// Verify the record was indexed by searching for it
        // This search should return records with "test" in their contents
	searchResp, err := client.Search(
```

Having said that, this is a personal preference. Comments can make the code look cluttered and increase scrolling, it can duplicate some of the post explanations, etc. I suggest adding them in the GitHub repo containing the demo code instead.

Author: Your comments are really great, and I think they will help readers. I  disseminated them through the blog post, adding them only on the first occurrence of code I duplicate, and adding more similar comments. I have also removed comments that are repeated when code is repeated, so that readers have an easier time noticing new code.

│Your service│          │ Wiremock in recording mode │       │ Algolia │
│            │◄─────────┤                            │◄──────┤         │
└────────────┘          └────────────────────────────┘       └─────────┘
```

In recording mode, you give Wiremock a URL to record, and it will store files
representing the requests you made, and the corresponding responses. With
Algolia, it can be quite long, especially if you [wait for
operations](https://www.algolia.com/doc/api-reference/api-methods/wait-task).
What happens in practice is that the SDK will use a polling mechanism to check
if your task is published. This will result in a lot of similarly looking files.
This is not very interesting to reproduce in your test, so I recommend simply
deleting files representing a negative response to the question: "are the
changes published yet?". Those typically contain a JSON field called `status`
set to `notPublished` in their body. The files have names that are
a bit ugly, so I usually rename them for clarity.
For example, you might rename
`1_indexes_test-index_task_226434943725-6e8689fa-9bbb-43fb-9d24-6824c02fc7d5.json`

> to `index_test_task_published.json`.

Reviewer 2: Maybe here having some Json content to illustrate the different status could make things a bit clearer

        1_indexes_test-index_task_226434943725                     |
                                                                   |
        GET                                                        | GET
        /1/indexes/test-index/task/226434943725                    | /1/indexes/test-index/task/226434943725
                                                                   |
        [Scenario                                                  | [Scenario                                           <<<<< Scenario does not match
        'scenario-1-1-indexes-test-index-task-226434943725'        | 'scenario-1-1-indexes-test-index-task-226434943725'
        state:                                                     | state: Started]
        scenario-1-1-indexes-test-index-task-226434943725-43]      |
                                                                   |
        -----------------------------------------------------------------------------------------------------------------------
```

After dropping `"requiredScenarioState" : "scenario-1-1-indexes-test-index-task-226434943725-43",`
from the mapping file about polling, the test passes again, only this time, it

> passes in under 2 seconds.

Reviewer 2: I think here we would benefit from a whole section about scenarios and how Wiremock handle it as it is a core thing for stateful/replay mechanism

Author: The blogpost is already a bit long and besides, I'm still very unfamiliar with Wiremock and its concepts like scenarios, so I won't venture in this territory for now.

Reviewer 2: Yes but as the replay feature is based on the stateful behaviour I think it is worth mentioning quickly this powerful: 

> It is possible for request to mention a scenario they belong to, allowing to do things like "On first 2 calls respond A and on 3rd return B".  Based on that it is possible to build complete choreography of request/response

Author: Thanks, added!

the following message:

```
tc-wiremock.go:73: create container: container create: Error response from daemon: Invalid bind mount config: mount source "/builds/product-discovery/ms.indexer/internal/import/brandsuggestion/testdata" is forbidden by the allow list [/home /tmp] - update the bind mounts configuration and restart the agent to enable
```

It would seem that we cannot use a bind mount in the CI. Let us use our
`record` constant to make the container options conditional:

```go
// indexer_test.go

if record {
	// Use a bind mount
	opts = append(opts, testcontainers.WithHostConfigModifier(func(hostConfig *container.HostConfig) {
		hostConfig.Binds = []string{
			absolutePath + "/testdata:/home/wiremock/mappings",
		}
	}))
} else {
	// Use a copy operation
	mappingFiles, err := os.ReadDir(absolutePath + "/testdata")
	if err != nil {
		t.Fatalf("Failed to read testdata directory: %v", err)
	}
	for _, mappingFile := range mappingFiles {
		opts = append(opts, testcontainers_wiremock.WithMappingFile(
			mappingFile.Name(),
			"testdata/"+mappingFile.Name(),
		))
	}
}
```

When recording, we mount the volume, which is not an issue because we are not
in the CI.

> Otherwise, we use the `WithMappingFile` function which relies on a copy operation.

Reviewer 2: An addition to testContainer Wiremock facade could be helpful here to explain why it is so easy to do, and that other things are also possible

Author: I'll add a paragraph about that, but I think I don't think I will mention the rest of the API exposed by this facade, as it is very small. There are not a lot of other helpful functions.

Reviewer 2: Ok maybe a link to do then.
My though was to add some information that is a provided APIs an not some dark magic 😄

Author: 👍 just added a link
